### PR TITLE
feat(derive_code_mappings): Use lock to get_trees_for_org()

### PR DESCRIPTION
**Pull request by armenzg**
For now, we will only add it to the dry_run path to test it out.

Fixes WOR-2466

### CODE REVIEW

File: src/sentry/tasks/derive_code_mappings.py [hunk 1]
```diff
@@ -1,24 +1,22 @@
...
 from sentry.models.integrations.repository_project_path_config import RepositoryProjectPathConfig
 from sentry.models.organization import Organization
 from sentry.models.repository import Repository
 from sentry.tasks.base import instrumented_task
-from sentry.utils.json import JSONData
+from sentry.utils.locking import UnableToAcquireLock
 from sentry.utils.safe import get_path
 
-ACTIVE_PROJECT_THRESHOLD = timedelta(days=7)
-GROUP_ANALYSIS_RANGE = timedelta(days=14)
```
Comment (armenzg) [inline src/sentry/tasks/derive_code_mappings.py:21]: Unused.

File: src/sentry/tasks/derive_code_mappings.py [hunk 2]
```diff
@@ -40,29 +40,48 @@ def derive_code_mappings(
     This task is queued at most once per hour per project, based on the ingested events.
     """
     project = Project.objects.get(id=project_id)
-    organization: Organization = Organization.objects.get(id=project.organization_id)
-    set_tag("organization.slug", organization.slug)
+    org: Organization = Organization.objects.get(id=project.organization_id)
+    set_tag("organization.slug", org.slug)
     # When you look at the performance page the user is a default column
-    set_user({"username": organization.slug})
+    set_user({"username": org.slug})
```
Comment (armenzg) [inline src/sentry/tasks/derive_code_mappings.py:46]: The last few lines is just some minor clean up

File: src/sentry/tasks/derive_code_mappings.py [hunk 3]
```diff
@@ -40,29 +40,48 @@ def derive_code_mappings(
...
+        "organization.slug": org.slug,
+    }
+    feat_key = "organizations:derive-code-mappings"
     # Check the feature flag again to ensure the feature is still enabled.
-    should_continue = features.has(
-        "organizations:derive-code-mappings", organization
-    ) or features.has("organizations:derive-code-mappings-dry-run", organization)
-    if not (dry_run or should_continue):
-        logger.info(f"Event from {organization.slug} org should not be processed.")
+    should_continue = features.has(feat_key, org) or features.has(f"{feat_key}-dry-run", org)
```
Comment (armenzg) [inline src/sentry/tasks/derive_code_mappings.py:53]: Collapsed into one line.

File: src/sentry/tasks/derive_code_mappings.py [hunk 4]
```diff
@@ -40,29 +40,48 @@ def derive_code_mappings(
...
 
-    installation, organization_integration = get_installation(organization)
+    installation, organization_integration = get_installation(org)
     if not installation:
         return
 
-    trees: JSONData = installation.get_trees_for_org()
+    trees = {}
+    if not dry_run:
+        trees = installation.get_trees_for_org()
```
Comment (armenzg) [inline src/sentry/tasks/derive_code_mappings.py:69]: Old behaviour until we verify that it works well.

File: src/sentry/tasks/derive_code_mappings.py [hunk 5]
```diff
@@ -40,29 +40,48 @@ def derive_code_mappings(
...
+    else:
+        # Acquire the lock for a maximum of 10 minutes
+        lock = locks.get(
+            key=f"get_trees_for_org:{org.slug}", duration=60 * 10, name="process_pending"
+        )
+
+        try:
+            with lock.acquire():
+                trees = installation.get_trees_for_org(3600 * 3)
```
Comment (armenzg) [inline src/sentry/tasks/derive_code_mappings.py:78]: The default caching is 24 hours, however, I want to see improvements quickly. I will revert on the next PR.

**Review by scttcper — APPROVED** [review]

**Review by snigdhas — APPROVED** [review]

File: src/sentry/tasks/derive_code_mappings.py [hunk 6]
```diff
@@ -40,29 +40,48 @@ def derive_code_mappings(
...
     # Check the feature flag again to ensure the feature is still enabled.
-    should_continue = features.has(
-        "organizations:derive-code-mappings", organization
-    ) or features.has("organizations:derive-code-mappings-dry-run", organization)
-    if not (dry_run or should_continue):
-        logger.info(f"Event from {organization.slug} org should not be processed.")
+    should_continue = features.has(feat_key, org) or features.has(f"{feat_key}-dry-run", org)
+
+    if not (dry_run or should_continue or data["platform"] not in SUPPORTED_LANGUAGES):
+        logger.info("Event should not be processed.", extra=extra)
```
[the remark this answers is not in the document]
Comment (markstory) [reply src/sentry/tasks/derive_code_mappings.py:56]: Is the org slug alone enough to help contextualize this log message?

**Review by markstory — APPROVED** [review]

**Review by wedamija — COMMENTED** [review]
Why do you want to retry if you can't acquire the lock here? You try to limit this to once per hour here: https://github.com/getsentry/sentry/blob/f2c1f3648cb2000f12a4025b2e509ff16fbc8280/src/sentry/tasks/post_process.py#L618-L620

Won't this counter that?

**Comment by wedamija** [comment]
> Why do you want to retry if you can't acquire the lock here? You try to limit this to once per hour here:
> 
> https://github.com/getsentry/sentry/blob/f2c1f3648cb2000f12a4025b2e509ff16fbc8280/src/sentry/tasks/post_process.py#L618-L620
> 
> Won't this counter that?

Discussed on slack, I missed that the lock is on the org level, not project 👍

**Pull request merged**